File: frontend/src/domain/word.ts
export const DIFFICULTIES = ['easy', 'medium', 'hard'] as const;

export type Difficulty = (typeof DIFFICULTIES)[number];

export function isDifficulty(value: string): value is Difficulty {
  return (DIFFICULTIES as readonly string[]).includes(value);
}

/** A word category, e.g. "Home", "Animals". Open set — defined by the seed data. */
export type Category = string;

export interface Word {
  id: number;
  english: string;
  arabic: string;
  category: Category;
  difficulty: Difficulty;
  enabled: boolean;
  /** Describer hint in English, stored per word. */
  hintEn: string;
  /** Describer hint in Arabic, stored per word. */
  hintAr: string;
}

export interface WordFilter {
  search?: string;
  difficulty?: Difficulty;
  category?: Category;
  enabledOnly?: boolean;
}

File: backend/internal/domain/word.go
package domain

import "time"

// Difficulty is one of the three fixed game difficulty levels.
// Values are lowercase by contract with the seed file and the frontend schema.
type Difficulty string

const (
	DifficultyEasy   Difficulty = "easy"
	DifficultyMedium Difficulty = "medium"
	DifficultyHard   Difficulty = "hard"
)

// Difficulties lists all valid difficulty levels in display order.
var Difficulties = []Difficulty{DifficultyEasy, DifficultyMedium, DifficultyHard}

// Valid reports whether d is one of the known difficulty levels.
func (d Difficulty) Valid() bool {
	switch d {
	case DifficultyEasy, DifficultyMedium, DifficultyHard:
		return true
	}
	return false
}

// Word is a single bilingual vocabulary card.
type Word struct {
	ID         int64
	English    string
	Arabic     string
	Category   string
	Difficulty Difficulty
	Enabled    bool
	HintEn     string // describer hint in English
	HintAr     string // describer hint in Arabic
	CreatedAt  time.Time
}
